src/adapter/src/catalog/migrate.rs

```diff
@@ -185,9 +185,28 @@ pub(crate) async fn migrate(
 
 /// Migrations that run only on the durable catalog before any data is loaded into memory.
 pub(crate) fn durable_migrate(
-    _tx: &mut Transaction,
+    tx: &mut Transaction,
     _boot_ts: Timestamp,
 ) -> Result<(), anyhow::Error> {
+    // We're promoting mz_cluster_replica_frontiers from mz_internal to
+    // mz_catalog in this release. As a special case, if a mapping for
+    // mz_cluster_replica_frontiers exists in the mz_internal schema, rewrite it
+    // in place for the mz_catalog schema instead. This keeps the migration
+    // machinery from creating a new shard for the relation, which is important
+    // because the 0dt machinery relies on this shard to determine when to cut
+    // over to the new version.
+    //
+    // TODO(benesch): remove this for the next release (27 September 2024).
+    let mut mappings = tx.get_system_object_mappings().collect::<Vec<_>>();
+    if let Some(object) = mappings.iter_mut().find(|o| o.description == SystemObjectDescription {
+        schema_name: "mz_internal".into(),
+        object_name: "mz_cluster_replica_frontiers".into(),
+        object_type: mz_sql::catalog::CatalogItemType::Source,
+    }) {
+        object.description.schema_name = "mz_catalog".into();
```

Comment:
Not a big deal, but these all have constants laying around, if they're accessible.

- `mz_repr::namespaces::MZ_INTERNAL_SCHEMA`
- `mz_repr::namespaces::MZ_CATALOG_SCHEMA`
- `mz_catalog::builtin::MZ_CLUSTER_REPLICA_FRONTIERS.name`